core/directory.go

```diff
@@ -239,6 +240,48 @@ func (dir *Directory) WithCopiedFile(ctx context.Context, subdir string, src *Fi
 	return NewDirectory(ctx, st, cwd, platform)
 }
 
+func MergeDirectories(ctx context.Context, dirs []*Directory, platform specs.Platform) (*Directory, error) {
+	states := make([]llb.State, 0, len(dirs))
+	for _, fs := range dirs {
+		state, _, dirPlatform, err := fs.Decode()
+		if err != nil {
+			return nil, err
+		}
+
+		if !reflect.DeepEqual(platform, dirPlatform) {
+			// TODO(vito): work around with llb.Copy shenangians?
+			return nil, fmt.Errorf("TODO: cannot diff across platforms: %+v != %+v", platform, dirPlatform)
+		}
+
+		states = append(states, state)
+	}
+	return NewDirectory(ctx, llb.Merge(states), "", platform)
+}
+
+func (dir *Directory) Diff(ctx context.Context, other *Directory) (*Directory, error) {
+	st, rel, platform, err := dir.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	otherSt, otherRel, otherPlatform, err := other.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	if rel != otherRel {
+		// TODO(vito): work around with llb.Copy shenangians?
+		return nil, fmt.Errorf("TODO: cannot diff with different relative paths: %q != %q", rel, otherRel)
+	}
+
+	if !reflect.DeepEqual(platform, otherPlatform) {
```

Comment:
It might work, I'm just not sure how to represent it, or test it. :thinking: I've only seen platform get passed to `Marshal` and wasn't sure what should happen when we have multiple states turn into one.

With `Diff` I guess it's simpler because it's really a one-directional diff, more like "what has upper changed from lower." But with `Merge` I have no idea what to keep. Maybe the first? Or last?

Should I just remove this check and have it carry on with `otherPlatform`? And probably call it `upper` instead of `other` for clarity?

Reply:
Ah okay right, you don't have choice but to choose a value, but I agree it's not clear which one is right. Honestly feel free to just leave this as is for now. It's not causing any harm and I will think about it some more when I rebase my multiplatform PR on the new API.